From: Ahmad Fatoum <a.fatoum@pengutronix.de>
Date: Mon, 11 Jul 2022 15:15:55 +0200
Subject: [PATCH] scripts: gen-dtb-s: fallback to LZ4 if lzop is unavailable

lzop hasn't seen any activity since 2017 and has been recently removed
from OpenEmbedded, which is unfortunate as we unconditonally use LZO for
compressing device trees that are referenced via __dtb_z_.

As a fallback, use LZ4 as compression if lzop is not available.

Signed-off-by: Ahmad Fatoum <a.fatoum@pengutronix.de>
---
 common/Kconfig    | 7 ++++++-
 scripts/gen-dtb-s | 6 ++++++
 2 files changed, 12 insertions(+), 1 deletion(-)

diff --git a/common/Kconfig b/common/Kconfig
index fa5314f7b15d..33afdcbb0725 100644
--- a/common/Kconfig
+++ b/common/Kconfig
@@ -1,5 +1,6 @@
 # SPDX-License-Identifier: GPL-2.0-only
 
+source "scripts/Kconfig.include"
 source "common/boards/Kconfig"
 
 config GREGORIAN_CALENDER
@@ -46,10 +47,14 @@ config USE_COMPRESSED_DTB
 	depends on ARM || RISCV
 	select UNCOMPRESS
 	select LZO_DECOMPRESS if DTB_COMPRESSION = "lzo"
+	select LZ4_DECOMPRESS if DTB_COMPRESSION = "lz4"
 
 config DTB_COMPRESSION
 	string
-	default "lzo"
+	default "none" if !USE_COMPRESSED_DTB
+	default "lzo" if $(success,lzop -V)
+	default "lz4" if $(success,lz4c -V)
+	default "none"
 
 config FILETYPE
 	bool
diff --git a/scripts/gen-dtb-s b/scripts/gen-dtb-s
index eda52aac1600..8331eb585743 100755
--- a/scripts/gen-dtb-s
+++ b/scripts/gen-dtb-s
@@ -59,6 +59,12 @@ case $compression in
 	lzo)
 		lzop -f -9 $dtb -o $dtb.$compression
 		;;
+	lz4)
+		lz4c -fql $dtb $dtb.$compression
+		# LZ4 decompressor strips away last 4 bytes, but doesn't
+		# consume them, so increase size accordingly
+		printf '\0\0\0\0' >>$dtb.$compression
+		;;
 	*)
 		echo Unsupported compression algorithm: $compression >&2
 		exit 1
